src/charm.py

```diff
@@ -631,29 +694,138 @@ def _set_leader_unit_active_if_needed(self):
+
+            # We keep the secret object around -- needed when applying modifications
+            self.secrets[scope][Config.Secrets.SECRET_LABEL] = secret
+
+            # We retrieve and cache actual secret data for the lifetime of the event scope
+            self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL] = secret.get_content()
+
+        return self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL]
+
+    def _juju_secret_get_key(self, scope: Scopes, key: str) -> Optional[str]:
+        if not key:
+            return
+
+        if self._juju_secrets_get(scope):
+            secret_cache = self.secrets[scope].get(Config.Secrets.SECRET_CACHE_LABEL)
+            if secret_cache:
+                secret_data = secret_cache.get(key)
+                if secret_data and secret_data != Config.Secrets.SECRET_DELETED_LABEL:
+                    logging.debug(f"Getting secret {scope}:{key}")
+                    return secret_data
+        logging.debug(f"No value found for secret {scope}:{key}")
+
+    def get_secret(self, scope: Scopes, key: str) -> Optional[str]:
+        """Getting a secret."""
+        peer_data = self._peer_data(scope)
+
+        juju_version = JujuVersion.from_environ()
+
+        if juju_version.has_secrets:
+            return self._juju_secret_get_key(scope, key)
+        else:
+            return peer_data.get(key)
+
+    def _juju_secret_set(self, scope: Scopes, key: str, value: str) -> str:
+        """Helper function setting Juju secret."""
+        peer_data = self._peer_data(scope)
+        self._juju_secrets_get(scope)
+
+        secret = self.secrets[scope].get(Config.Secrets.SECRET_LABEL)
+
+        # It's not the first secret for the scope, we can re-use the existing one
+        # that was fetched in the previous call
+        if secret:
+            secret_cache = self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL]
+
+            if secret_cache.get(key) == value:
+                logging.debug(f"Key {scope}:{key} has this value defined already")
+            else:
+                secret_cache[key] = value
+                try:
+                    secret.set_content(secret_cache)
+                except OSError as error:
+                    logging.error(
+                        f"Error in attempt to set {scope}:{key}. "
+                        f"Existing keys were: {list(secret_cache.keys())}. {error}"
+                    )
+                logging.debug(f"Secret {scope}:{key} was {key} set")
+
+        # We need to create a brand-new secret for this scope
+        else:
+            scope_obj = self._scope_opj(scope)
+
+            secret = scope_obj.add_secret({key: value})
+            if not secret:
+                raise SecretNotAddedError(f"Couldn't set secret {scope}:{key}")
+
+            self.secrets[scope][Config.Secrets.SECRET_LABEL] = secret
+            logging.debug(f"Secret {scope}:{key} published (as first). ID: {secret.id}")
```

Comment:
Hmmm, wait... 
Shouldn't we just rather do:
```
self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL] = {key: value}
```
here to avoid fetching? (Or did I get something wrong? `get_content()` should return us `{key: value}` at this point --assuming no stale read ;-)  Why to fetch at all, instead of using what's already available here?)

Reply:
yes, LGTM